code/datums/progress_bar.dm

```diff
@@ -0,0 +1,182 @@
+/**
+  * # Progress Bar
+  *
+  * Use to give feedback on the progress of an action that takes time.
+  * Can use a custom theme which alters the visuals of the bar.
+  *
+  * How does it work visually?
+  * There are 3 images comprising the progress bar:
+  * - [theme]_backdrop - The bar's background.
+  * - [theme]_mask - The mask, which hides the fill and decreases in size, exposing more of the fill as the bar progresses.
+  * - [theme]_fill - The bar's fill, ideally white so it can be coloured.
+  * The result is there is no longer a need to create states for every 5% - the filling is instead interpolated and colored accordingly.
+  */
+/datum/progress_bar
+	// Settings
+	/// The goal value to reach. When [/datum/progress_bar/proc/update] is called, a fraction is calculated with the goal.
+	var/goal = 100
+	/// The icon file to use for the visual element.
+	var/icon_path = 'icons/effects/progress_bar.dmi'
+	/// The progress bar visual theme.
+	var/theme = "default"
+	/// The total height in pixels of the theme's progress bar.
+	var/height = 7
+	/// The width in pixels of the theme's fill sprite.
+	var/fill_width = 22
+	// Variables
+	/// The target to display the progress bar on.
+	var/atom/target
+	/// The main visual element holding everything up.
+	var/image/holder
+	/// The backdrop visual element. Contains the fill and mask as overlays.
+	var/image/backdrop
+	/// The fill visual element.
+	var/image/fill
+	/// The mob currently owning the progress bar, used to calculate the bar's vertical position.
+	var/mob/user
+	/// The client currently owning the progress bar for display.
+	var/client/client
+	/// The bar's index number within the owning mob's active progress bar list.
+	var/list_index
+	/// Whether the bar is stopping.
+	var/stopping = FALSE
+
+/datum/progress_bar/New(mob/user, goal, atom/target)
+	. = ..()
+	if(!istype(target))
+		EXCEPTION("Invalid target given")
```

Comment:
I don't think this even throws the exception. Probably best to use `CRASH`, unless you have further handling for invalid targets below.